### PR TITLE
Lambda functions depends on transformed role

### PR DESCRIPTION
**Pull request by justcompile**
*Issue #, if available:*  n/a

*Description of changes:*

We believe that we found a bug where by using Fn::Sub with a lambda function's `FunctionName` would cause the function creation to fail due to the Role not being available despite it existing in IAM & Cloudformation reporting that it had created the role successfully.

This change ensures that the transformed Function resource depends on the automatically created IAM role.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by brettstack** [comment]
Hi @justcompile, could you provide a reproduceable template so that we can confirm the behavior? I'm hesitant to take in a change which affects every test.

**Comment by justcompile** [comment]
Hi @brettstack, we've had to work around it but I believe the following should be enough to confirm (minus the actual function code).

If not, let me know
```
AWSTemplateFormatVersion: '2010-09-09'
Transform: 'AWS::Serverless-2016-10-31'
Description: >-
  Integration Data Pipeline 2.
Parameters:
  NamePrefix:
    Description: Used to form a convention to prefix generated entities
    Type: String
  EnvironmentName:
    Type: String

Globals:

  Api:
    # Enable CORS for API Gateway endpoints proxying to the Lambda functions.
    Cors:
      AllowMethods: "'POST,OPTIONS'"
      AllowHeaders: "'Content-Type'"
      AllowOrigin: "'http://localhost:3000'"

Resources:
  LambdaFunctionSecurityGroup:
    Type: AWS::EC2::SecurityGroup
    Properties:
      GroupName: !Sub "${NamePrefix}-Lambda-Egress"
      GroupDescription: !Sub "${NamePrefix} Lambda Internet Access"
      SecurityGroupEgress:
        - IpProtocol: tcp
          FromPort: 80
          ToPort: 80
          CidrIp: 0.0.0.0/0
        - IpProtocol: tcp
          FromPort: 443
          ToPort: 443
          CidrIp: 0.0.0.0/0
      Tags:
        -
          Key: "Team"
          Value: "io"
        -
          Key: "Environment"
          Value: !Ref EnvironmentName
        -
          Key: "Prefix"
          Value: !Ref NamePrefix
      VpcId:
        Fn::ImportValue: !Sub io-${EnvironmentName}-VPC

  IntegrationsDirector:
    DependsOn:
      - LambdaFunctionSecurityGroup
    Type: 'AWS::Serverless::Function'
    Properties:
      FunctionName: !Sub "${NamePrefix}-director"
      Handler: director.handler.lambda_handler
      Runtime: python3.6
      CodeUri: ../project/
      Description: >-
      The Director examines a jobspec in its director bucket and
      activates the appropriate state machine for the integration
      the jobspec is intended for.
      MemorySize: 128
      Timeout: 30
      Tracing: Active
      Policies:
      - VPCAccessPolicy: {}
      - arn:aws:iam::aws:policy/service-role/AWSLambdaVPCAccessExecutionRole
      - Version: '2012-10-17'
          Statement:
          - Effect: Allow
            Action:
              - "states:StartExecution"
              - "states:ListStateMachines"
              - "states:ListExecutions"
            Resource: '*'
      Events:
        CreateJob:
          Type: Api
          Properties:
            Path: /sync
            Method: post
      Environment:
        Variables:
          PYTHONPATH: "/var/runtime:/var/task/:/var/task/dist"
          BASE_STACK_NAME: !Sub '${NamePrefix}'
      VpcConfig:
        SecurityGroupIds:
          - !GetAtt LambdaFunctionSecurityGroup.GroupId
        SubnetIds:
          Fn::Split:
          - ","
          - Fn::ImportValue: !Sub 'io-${EnvironmentName}-PrivateSubnets'
```

**Comment by brettstack** [comment]
Thanks. I think I understand the problem, though I will try to reproduce myself to make sure. I'm a little confused why it's causing issue since you said "IAM & Cloudformation reporting that it had created the role successfully". Does that get created before it attempts to create the Lambda function? What's the exact error/reason you're seeing for failure in CloudFormation?

**Comment by justcompile** [comment]
I can't grab the exact error as we've added a work around, but the error we were seeing was along the lines of "Unable to create Function, role <Role name> does not exist". I believe it is down to a race condition as the function does not have a "dependson" property, so when SAM creates a role based on the Policies property, it could be created before, after or during the function creation

File: samtranslator/model/sam_resources.py [hunk 1]
```diff
@@ -115,6 +115,10 @@ def to_cloudformation(self, **kwargs):
             execution_role = self._construct_role(managed_policy_map)
             lambda_function.Role = execution_role.get_runtime_attr('arn')
             resources.append(execution_role)
+            try:
```
[the remark this answers is not in the document]
Comment (justcompile) [reply samtranslator/model/sam_resources.py]: Of course. I'm used to the EAFP coding style (https://docs.python.org/3.4/glossary.html) for Python, but happy to change it

**Comment by jlhood** [comment]
@justcompile Hi, thanks so much for taking the time to submit this PR. However, I don't think it's necessary to create an explicit dependency between the `AWS::Lambda::Function` resource and the `AWS::IAM::Role` resource. CloudFormation understands implicit dependencies when a property of one resource is a reference (or GetAtt) to another resource. When SAM transforms `AWS::Serverless::Function` into an `AWS::Lambda::Function` and `AWS::IAM::Role`, the `AWS::Lambda::Function` Role property is populated by doing a `!GetAtt` on the `AWS::IAM::Role` to get the ARN. CloudFormation understands this dependency and will always create the IAM Role first before creating the Lambda function (it has to in order to resolve the `!GetAtt`).

So I think the issue you were experiencing where AWS Lambda was failing saying the role was not found was another issue. I looked in the history of the [AWS Service Health Dashboard](https://status.aws.amazon.com/) and found that IAM had posted an informational message on October 25 (2 days before you opened this PR):

> 10:40 AM PDT We continue to investigate increased latency on administrative APIs. Create, Delete, List, Get, and Update API actions may be impacted in multiple regions. Other AWS services like AWS CloudFormation and AWS Lambda that use IAM roles may also be impacted. User authentications and authorizations are not impacted.

So what you experienced could have been from the above event.

**Pull request closed without merging**